tests/Doctrine/Tests/Common/Annotations/LegacyCachedReaderTest.php

```diff
@@ -0,0 +1,165 @@
+     * @group 62
+     * @group 105
+     */
+    public function testUsesFreshCacheWithTraitsThatUseOtherTraits()
+    {
+        $cacheTime = time();
+
+        touch(__DIR__ . '/Fixtures/ClassThatUsesTraitThatUsesAnotherTrait.php', $cacheTime - 10);
+        touch(__DIR__ . '/Fixtures/Traits/EmptyTrait.php', $cacheTime - 10);
+
+        $this->doTestCacheFresh(
+            'Doctrine\Tests\Common\Annotations\Fixtures\ClassThatUsesTraitThatUsesAnotherTrait',
+            $cacheTime
+        );
+    }
+
+    protected function doTestCacheStale($className, $lastCacheModification)
+    {
+        $cacheKey       = strtr($className, '\\', '.');
+
+        /* @var $cache CacheProvider|\PHPUnit_Framework_MockObject_MockObject */
+        $cache = $this->createMock(CacheProvider::class);
+        $cache
+            ->expects($this->at(0))
+            ->method('fetch')
+            ->with($this->equalTo($cacheKey))
+            ->will($this->returnValue([])) // Result was cached, but there was no annotation
+        ;
+        $cache
+            ->expects($this->at(1))
+            ->method('fetch')
+            ->with($this->equalTo('[C]'.$cacheKey))
+            ->will($this->returnValue($lastCacheModification))
+        ;
+        $cache
+            ->expects($this->at(2))
+            ->method('save')
+            ->with($this->equalTo($cacheKey))
+            ->willReturn(true)
+        ;
+        $cache
+            ->expects($this->at(3))
+            ->method('save')
+            ->with($this->equalTo('[C]'.$cacheKey))
+            ->willReturn(true)
+        ;
+
+        $reader = new CachedReader(new AnnotationReader(), $cache, true);
+        $route = new Route();
+        $route->pattern = '/someprefix';
+
+        self::assertEquals([$route], $reader->getClassAnnotations(new \ReflectionClass($className)));
+    }
+
+    protected function doTestCacheFresh($className, $lastCacheModification)
+    {
+        $cacheKey       = strtr($className, '\\', '.');
+        $route          = new Route();
+        $route->pattern = '/someprefix';
+
+        /* @var $cache CacheProvider|\PHPUnit_Framework_MockObject_MockObject */
+        $cache = $this->createMock(CacheProvider::class);
+        $cache
+            ->expects($this->at(0))
+            ->method('fetch')
+            ->with($this->equalTo($cacheKey))
+            ->will($this->returnValue([$route])); // Result was cached, but there was an annotation;
+        $cache
+            ->expects($this->at(1))
+            ->method('fetch')
+            ->with($this->equalTo('[C]' . $cacheKey))
+            ->will($this->returnValue($lastCacheModification));
+        $cache->expects(self::never())->method('save');
+
+        $reader = new CachedReader(new AnnotationReader(), $cache, true);
+
+        $this->assertEquals([$route], $reader->getClassAnnotations(new \ReflectionClass($className)));
+    }
+
+    protected function getReader()
+    {
+        $this->cache = new ArrayCache();
```

Comment:
Do we also have a test running the CachedReader against an actual PSR-16 implementation, or only against the mock ? IMO, having a functional test of the CachedReader would be important (as PSR-16 implementations enforce some rules on the cache keys for instance).

This legacy test actually performs such a functional test for now (as the actual doctrine/cache cache will be wrapped by the actual PSR-16 adapter), but this legacy test is meant to be removed in the future, and we should not loose that functional testing.

Reply:
OK, looking at it again, CachedReader does not always use mocks, so it already fits it.

Reply:
Yes we do, I copy/pasted/adapted this file to ` tests/Doctrine/Tests/Common/Annotations/LegacyCachedReaderTest.php`